### PR TITLE
Migrating to Node 16

### PR DESCRIPTION
**Pull request by AndreyIvanov42**
## Task description

In order to migrate tasks to the Node16 execution handler, we have to update the pipelines-task-lib itself to support the Node16 runtime.

## Changelog

- Rewrote the pipeline to use node 16
- Updated dev dependencies
- Updated package-lock.json to version 2 to unlock the ability for deterministic and reproducible builds
- Fixed types for tests
- Fixed tool runner test to utilize changed output from Node 16
- Fixed `decipher.update` method call

## Testing

- Changes have been tested on Linux, MacOS, Windows: 
  by unit-tests
  manually as part of some task. Used tasks and cases are described [here](https://github.com/microsoft/azure-pipelines-task-lib/pull/844#issuecomment-1169891938)
- Enough test coverage for changes and current test coverage for the task doesn't look poor
- I didn't find any breaking changes
- There are no other concerns
- I haven't discovered any new uncovered test/use cases

Added unit tests: No

### CODE REVIEW

**Review by DenisRumyantsev — APPROVED** [review]

**Review by kirill-ivlev — APPROVED** [review]

File: node/vault.ts [hunk 1]
```diff
@@ -67,7 +67,7 @@ export class Vault {
             var encryptedText = ivDataBuffer.slice(16);
 
             var decipher = crypto.createDecipheriv(algorithm, key, iv);
-            var dec = decipher.update(encryptedText,encryptEncoding,unencryptedEncoding);
+            var dec = decipher.update(encryptedText);
```
Comment (max-zaytsev) [inline node/vault.ts:70]: Just curious why are we removing this?

File: node/package-lock.json [hunk 2]
```diff
@@ -1,8 +1,1391 @@
 {
   "name": "azure-pipelines-task-lib",
-  "version": "4.0.0-preview",
-  "lockfileVersion": 1,
+  "version": "4.0.0-preview.0",
+  "lockfileVersion": 2,
```
Comment (alexander-smolyakov) [inline node/package-lock.json:4]: Could you please clarify why we have to change the `lockfileVersion` version?

File: azure-pipelines.yml [hunk 3]
```diff
@@ -70,20 +70,20 @@ jobs:
...
+  # use node 16
   - task: NodeTool@0
-    displayName: (azure-pipelines-task-lib) use node 10.23.0
+    displayName: (azure-pipelines-task-lib) use node 16.13.0
     inputs:
-      versionSpec: "10.23.0"
-      
+      versionSpec: "16.13.0"
+
   # build/test
```
Comment (alexander-smolyakov) [inline azure-pipelines.yml:87]: Could you please add a `PublishPipelineArtifact` step to publish the SDK build as a pipeline artifact?

Also, I believe we can extract common pipeline steps to the template.

**Review by alexander-smolyakov — COMMENTED** [review]
Please take a look at the comments below.

Also, please add a contribution section to the `README.md` file in the follow-up pull request to describe the current workflow:
- `master` branch for SDK v3 (stable and targets Node10)
- `releases/4.x` branch for SDK v4 (preview and targets Node16)
- Changes that were merged for SDK v3 should be cherry-picked for SDK v4 once they are merged to the `master branch`

File: node/package-lock.json (hunk 2 above)
Comment (AndreyIvanov42) [reply node/package-lock.json:4]: Because version 2 is the standard version for npm that comes with Node 16

**Review by alexander-smolyakov — APPROVED** [review]
LGTM, thanks!

**Review by wojgasior — APPROVED** [review]

**Pull request merged**